compiler/passes/denormalize.cpp

```diff
@@ -0,0 +1,602 @@
+ * 
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * 
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "astutil.h"
+#include "baseAST.h"
+#include "stlUtil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "symbol.h"
+#include "type.h"
+#include "LoopStmt.h"
+#include "CForLoop.h"
+#include "WhileStmt.h"
+
+//helper datastructures/types
+
+//maybe more memoization than caching
+std::map<Expr*,bool> safeExprCache;
+std::map<FnSymbol*,bool> globalManipFuncCache;
+
+typedef std::map<SymExpr*, std::pair<Expr*,Type*> > ActualUseDefCastMap;
+
+//prototypes
+bool isExprSafeForReorder(Expr * e);
+bool canPrimMoveCreateCommunication(CallExpr* ce);
+inline bool possibleDepInBetween(Expr* e1, Expr* e2);
+inline bool requiresCast(Type* t);
+inline bool isIntegerPromotionPrimitive(PrimitiveTag tag);
+inline bool isNonEssentialPrimitive(CallExpr* ce);
+bool isSafePrimitive(CallExpr* ce);
+
+bool isDenormalizable(Symbol* sym,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap, SymExpr** useOut, Expr** defOut,
+    Type** castTo);
+
+void denormalizeActuals(CallExpr* ce,
+    Map<Symbol*,Vec<SymExpr*>*> defMap,
+    Map<Symbol*,Vec<SymExpr*>*> useMap,
+    ActualUseDefCastMap actualUseDefMap);
+
+void denormalize(void);
+void denormalize(FnSymbol *fn);
+void denormalize(Expr* def, SymExpr* use, Type* castTo);
+
+
+void denormalize(void) {
+  if(fDenormalize) {
+    forv_Vec(FnSymbol, fn, gFnSymbols) {
+      denormalize(fn);
+    }
+  }
+}
+
+void denormalize(FnSymbol *fn) {
+
+  Vec<Symbol*> symSet;
+  Vec<SymExpr*> symExprs;
+  Map<Symbol*,Vec<SymExpr*>*> defMap;
+  Map<Symbol*,Vec<SymExpr*>*> useMap;
+
+  collectSymbolSetSymExprVec(fn, symSet, symExprs);
+  buildDefUseMaps(symSet, symExprs, defMap, useMap);
+
+  //data structures for deferred actual handling
+  std::set<CallExpr*> deferredFns;
+  ActualUseDefCastMap actualUseDefMap;
+
+  bool cachedGlobalManip = globalManipFuncCache.count(fn);
+
+  forv_Vec(Symbol, sym, symSet) {
+
+    SymExpr *use = NULL;
+    Expr *usePar = NULL;
+    Expr *def = NULL;
+    Expr *defPar;
+    Type* castTo = NULL;
+
+    //if we don't already have it cached,
+    //check for global symbols in function body
+    if(!cachedGlobalManip) {
+      //I am not sure if there should be !isConstant and !isParam as well
+      //I think there can be a const but stateful global that is manipulated by
+      //the function so, play safe
+      if(sym && !sym->isImmediate() && isGlobal(sym)){
```

Comment:
I don't understand the purpose of globalManipFuncCache. I'd expect you to maintain a map from FnSymbo* -> bool where the bool indicates whether or not a call to that function could modify state (and so should prevent a reordering as is possible with denormalization). I'd like to see that analysis written in a separate C file, even if for now it only works at this point in compilation. I'd personally call it "side effect free" or "pure". I'd expect the code to do this side-effect-free analysis separately from the denormalization, where they seem unnecessarily and confusingly coupled now. (But I could be misunderstanding).